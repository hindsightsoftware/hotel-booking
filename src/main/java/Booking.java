import java.util.Date;

public class Booking {

    private String firstname;
    private String lastname;
    private int price;
    private boolean booked;
    private BookingDates bookingDates;
    private String additional;

    public Booking(String firstname, String lastname, int price, boolean booked, Date checkin, Date checkout, String additional) {
        this.firstname = firstname;
        this.lastname = lastname;
        this.price = price;
        this.booked = booked;
        this.bookingDates = new BookingDates(checkin, checkout);
        this.additional = additional;
    }

    @Override
    public String toString() {
        return "Booking{" +
                "firstname='" + firstname + '\'' +
                "\n, lastname='" + lastname + '\'' +
                "\n, price=" + price +
                "\n, booked=" + booked +
                "\n, bookingDates=" + bookingDates.toString() +
                "\n, additional='" + additional + '\'' +
                "\n}";
    }
}
