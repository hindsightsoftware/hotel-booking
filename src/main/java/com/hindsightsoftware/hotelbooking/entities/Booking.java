package com.hindsightsoftware.hotelbooking.entities;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import java.util.Date;

@Entity
public class Booking {
    @Id
    @GeneratedValue(strategy=GenerationType.AUTO)
    private Long id;
    private String firstName;
    private String lastName;
    private int totalPrice;
    private boolean depositPaid;
    private Date checkIn;
    private Date checkOut;
    private String additionalNeeds;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public int getTotalPrice() {
        return totalPrice;
    }

    public void setTotalPrice(int totalPrice) {
        this.totalPrice = totalPrice;
    }

    public boolean isDepositPaid() {
        return depositPaid;
    }

    public void setDepositPaid(boolean depositPaid) {
        this.depositPaid = depositPaid;
    }

    public Date getCheckIn() {
        return checkIn;
    }

    public void setCheckIn(Date checkIn) {
        this.checkIn = checkIn;
    }

    public Date getCheckOut() {
        return checkOut;
    }

    public void setCheckOut(Date checkOut) {
        this.checkOut = checkOut;
    }

    public String getAdditionalNeeds() {
        return additionalNeeds;
    }

    public void setAdditionalNeeds(String additionalNeeds) {
        this.additionalNeeds = additionalNeeds;
    }
}
