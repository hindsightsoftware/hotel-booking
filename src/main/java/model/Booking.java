package model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Date;

public class Booking {

    @JsonProperty
    private String firstname;
    @JsonProperty
    private String lastname;
    @JsonProperty
    private int totalprice;
    @JsonProperty
    private boolean depositpaid;
    @JsonProperty(value = "bookingdates")
    private BookingDates bookingDates;
    @JsonProperty
    private String additionalneeds;

    public Booking(String firstname, String lastname, int totalprice, boolean depositpaid, Date checkin, Date checkout, String additionalneeds) {
        this.firstname = firstname;
        this.lastname = lastname;
        this.totalprice = totalprice;
        this.depositpaid = depositpaid;
        this.bookingDates = new BookingDates(checkin, checkout);
        this.additionalneeds = additionalneeds;
    }

    public Booking() {
    }

    public String getFirstname() {
        return firstname;
    }

    public String getLastname() {
        return lastname;
    }

    public int getTotalprice() {
        return totalprice;
    }

    public boolean isDepositpaid() {
        return depositpaid;
    }

    public BookingDates getBookingDates() {
        return bookingDates;
    }

    public String getAdditionalneeds() {
        return additionalneeds;
    }

    public void setFirstname(String firstname) {
        this.firstname = firstname;
    }

    public void setLastname(String lastname) {
        this.lastname = lastname;
    }

    public void setTotalprice(int totalprice) {
        this.totalprice = totalprice;
    }

    public void setDepositpaid(boolean depositpaid) {
        this.depositpaid = depositpaid;
    }

    public void setBookingDates(BookingDates bookingDates) {
        this.bookingDates = bookingDates;
    }

    public void setAdditionalneeds(String additionalneeds) {
        this.additionalneeds = additionalneeds;
    }

    @Override
    public String toString() {
        return "model.Booking{" +
                "firstname='" + firstname + '\'' +
                "\n, lastname='" + lastname + '\'' +
                "\n, totalprice=" + totalprice +
                "\n, depositpaid=" + depositpaid +
                "\n, bookingDates=" + bookingDates.toString() +
                "\n, additionalneeds='" + additionalneeds + '\'' +
                "\n}";
    }
}
