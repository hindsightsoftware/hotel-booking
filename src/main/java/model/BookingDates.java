package model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Date;

public class BookingDates {

    @JsonProperty
    private Date checkin;
    @JsonProperty
    private Date checkout;

    public BookingDates(Date checkin, Date checkout) {
        this.checkin = checkin;
        this.checkout = checkout;
    }

    public Date getCheckin() {
        return checkin;
    }

    public Date getCheckout() {
        return checkout;
    }

    @Override
    public String toString() {
        return "\nmodel.BookingDates{" +
                "\ncheckin=" + checkin +
                "\n, checkout=" + checkout +
                '}';
    }
}
