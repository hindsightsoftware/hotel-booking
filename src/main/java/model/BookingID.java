package model;

import com.fasterxml.jackson.annotation.JsonProperty;

public class BookingID {

    @JsonProperty
    private int id;

    public BookingID(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    @Override
    public String toString() {
        return "BookingID{" +
                "id=" + id +
                '}';
    }
}
