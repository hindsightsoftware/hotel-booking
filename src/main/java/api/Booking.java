package api;

import model.BookingModel;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.util.Date;

@RestController
public class Booking {

    @RequestMapping(value = "/booking", method = RequestMethod.GET)
    public BookingModel returnBooking() {
        Date checkin = new Date();
        checkin.setTime(1514764800);
        Date checkout = new Date();
        checkout.setTime(1514851200);

        BookingModel booking = new BookingModel("Mark", "Winteringham", 111, true, checkin, checkout, "Breakfast");

        return booking;
    }

}
