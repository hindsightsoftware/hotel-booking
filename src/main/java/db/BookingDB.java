package db;

import model.Booking;
import model.CreatedBooking;
import org.h2.jdbcx.JdbcDataSource;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.SimpleDateFormat;

public class BookingDB {

    private Connection conn;

    public BookingDB() throws SQLException {
        JdbcDataSource ds = new JdbcDataSource();
        ds.setURL("jdbc:h2:" + System.getProperty("user.dir") + "/booking.db");
        ds.setUser("sa");
        ds.setPassword("sa");
        conn = ds.getConnection();
    }

    public CreatedBooking create(Booking booking) throws SQLException {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");

        String sql = "INSERT INTO bookings(firstname, lastname, totalprice, deposit, checkin, checkout, additional) VALUES("
                     + "'" + booking.getFirstname() + "',"
                     + "'" + booking.getLastname() + "',"
                     + booking.getTotalprice() + ","
                     + booking.isDepositpaid() + ","
                     + "'" + dateFormat.format(booking.getBookingDates().getCheckin()) + "',"
                     + "'" + dateFormat.format(booking.getBookingDates().getCheckout()) + "',"
                     + "'" + booking.getAdditionalneeds() + "'"
                     + ")";

        int updateSuccess = conn.prepareStatement(sql).executeUpdate();

        if(updateSuccess > 0){
            ResultSet lastInsertId = conn.prepareStatement("SELECT LAST_INSERT_ID()").executeQuery();
            lastInsertId.next();

            String querySql = "SELECT * FROM bookings WHERE id='" + lastInsertId.getInt("LAST_INSERT_ID()") + "'";

            ResultSet result = conn.prepareStatement(querySql).executeQuery();
            result.next();

            Booking createdBooking = new Booking.BookingBuilder()
                                            .setFirstname(result.getString("firstname"))
                                            .setLastname(result.getString("lastname"))
                                            .setTotalprice(result.getInt("totalprice"))
                                            .setDepositpaid(result.getBoolean("deposit"))
                                            .setCheckin(result.getDate("checkin"))
                                            .setCheckout(result.getDate("checkout"))
                                            .setAdditionalneeds(result.getString("additional"))
                                            .build();

            return new CreatedBooking(result.getInt("id"), booking);
        } else {
            return null;
        }
    }

    public Booking query(int id) throws SQLException{
        String sql = "SELECT * FROM bookings WHERE id='" + id + "'";

        ResultSet result = conn.prepareStatement(sql).executeQuery();
        result.next();

        return new Booking.BookingBuilder()
                          .setFirstname(result.getString("firstname"))
                          .setLastname(result.getString("lastname"))
                          .setTotalprice(result.getInt("totalprice"))
                          .setDepositpaid(result.getBoolean("deposit"))
                          .setCheckin(result.getDate("checkin"))
                          .setCheckout(result.getDate("checkout"))
                          .setAdditionalneeds(result.getString("additional"))
                          .build();
    }

    public Boolean delete(int bookingid) throws SQLException {
        String sql = "DELETE FROM bookings WHERE id='" + bookingid + "'";

        int resultSet = conn.prepareStatement(sql).executeUpdate();
        System.out.println(resultSet);
        return resultSet == 1;
    }
}
