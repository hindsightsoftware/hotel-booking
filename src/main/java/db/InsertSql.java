package db;

import model.Booking;

import java.text.SimpleDateFormat;
import java.util.Date;

public class InsertSql {

    private String firstname;
    private String lastname;
    private int totalprice;
    private boolean depositpaid;
    private Date checkin;
    private Date checkout;
    private String additionalneeds;
    private SimpleDateFormat dateFormat;

    InsertSql(Booking booking) {
        this.firstname = booking.getFirstname();
        this.lastname = booking.getLastname();
        this.totalprice = booking.getTotalprice();
        this.depositpaid = booking.isDepositpaid();
        this.checkin = booking.getBookingDates().getCheckin();
        this.checkout = booking.getBookingDates().getCheckout();
        this.additionalneeds = booking.getAdditionalneeds();

        dateFormat = new SimpleDateFormat("yyyy-MM-dd");
    }

    public String buildSql(){
        return "INSERT INTO bookings(firstname, lastname, totalprice, deposit, checkin, checkout, additional) VALUES(" +
               "'" + firstname + "'," +
               "'" + lastname + "'," +
               totalprice + "," +
               depositpaid + "," +
               "'" + dateFormat.format(checkin) + "'," +
               "'" + dateFormat.format(checkout) + "'," +
                "'" + additionalneeds + "',);";
    }

}
