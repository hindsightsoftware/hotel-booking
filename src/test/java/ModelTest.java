import db.BookingDB;
import model.Booking;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.sql.SQLException;
import java.util.Date;

public class ModelTest {

    private Date checkin = new Date();
    private Date checkout = new Date();

    @Test
    public void testModel(){
        checkin.setTime(1514764800);
        checkout.setTime(1514851200);

        Booking booking = new Booking("Mark", "Winteringham", 111, true, checkin, checkout, "Breakfast");

        Approvals.verify(booking.toString());
    }

    @Test
    public void testAddBooking() throws SQLException {
        checkin.setTime(1514764800);
        checkout.setTime(1514851200);

        Booking booking = new Booking("Mark", "Winteringham", 111, true, checkin, checkout, "Breakfast");

        BookingDB bookingDB = new BookingDB();
        int result = bookingDB.create(booking);

        Approvals.verify(result);
    }

}
