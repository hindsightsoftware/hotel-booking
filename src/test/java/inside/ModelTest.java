package inside;

import db.BookingDB;
import model.Booking;
import model.CreatedBooking;
import org.approvaltests.Approvals;
import org.h2.jdbcx.JdbcDataSource;
import org.junit.Before;
import org.junit.Test;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Date;

public class ModelTest {

    @Before
    public void resetDatabase() throws SQLException {
        JdbcDataSource ds = new JdbcDataSource();
        ds.setURL("jdbc:h2:/Users/mark/Documents/restful-booker-java/booking.db");
        ds.setUser("sa");
        ds.setPassword("sa");
        Connection conn = ds.getConnection();

        String sql = "ALTER TABLE bookings ALTER COLUMN id RESTART WITH 1";
        conn.prepareStatement(sql).execute();
    }

    private Date checkin = new Date();
    private Date checkout = new Date();

    @Test
    public void testAddBooking() throws SQLException {
        checkin.setTime(1514764800);
        checkout.setTime(1514851200);

        Booking booking = new Booking.BookingBuilder()
                                     .setFirstname("Mark")
                                     .setLastname("Winteringham")
                                     .setTotalprice(123)
                                     .setDepositpaid(true)
                                     .setCheckin(checkin)
                                     .setCheckout(checkout)
                                     .setAdditionalneeds("Breakfast")
                                     .build();

        BookingDB bookingDB = new BookingDB();
        CreatedBooking result = bookingDB.create(booking);

        Approvals.verify(result);
    }

    @Test
    public void testGetBooking() throws SQLException {
        checkin.setTime(1514764800);
        checkout.setTime(1514851200);

        BookingDB bookingDB = new BookingDB();
        Booking result = bookingDB.query(1);

        Approvals.verify(result);
    }

}
