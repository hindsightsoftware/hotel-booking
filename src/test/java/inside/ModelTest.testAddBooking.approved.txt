CreatedBooking{
bookingid=1
, booking=model.Booking{firstname='Mark'
, lastname='Winteringham'
, totalprice=123
, depositpaid=true
, bookingDates=
model.BookingDates{
checkin=Sun Jan 18 13:46:04 GMT 1970
, checkout=Sun Jan 18 13:47:31 GMT 1970}
, additionalneeds='Breakfast'
}}
